### PR TITLE
Track Tasks State

### PR DESCRIPTION
**Pull request by dadgar**
This PR:

* Adds TaskState to the allocation.
* The Client is updated to track the Task through the States: Pending, Running, Dead and records the events that caused state transitions between these.

On a side note, this PR also fixes the alloc_runner and task_runner test regressions introduced with restart policies.

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -110,3 +110,29 @@ func (t *Task) Constrain(c *Constraint) *Task {
 	t.Constraints = append(t.Constraints, c)
 	return t
 }
+
+// TaskState tracks the current state of a task and events that caused state
+// transistions.
+type TaskState struct {
+	State  string
+	Events []*TaskEvent
```
Comment (diptanu) [inline api/tasks.go]: Are we bounding the size of Events somewhere? A service style job might potentially have a lot of events over the course of time if the node on which it is placed remains healthy for a really long time.

File: client/alloc_runner.go [hunk 2]
```diff
@@ -98,20 +98,22 @@ func (r *AllocRunner) RestoreState() error {
...
 		restartTracker := newRestartTracker(r.alloc.Job.Type, r.RestartPolicy)
-		tr := NewTaskRunner(r.logger, r.config, r.setTaskStatus, r.ctx, r.alloc.ID, task, restartTracker)
+		tr := NewTaskRunner(r.logger, r.config, r.setTaskState, r.ctx,
+			r.alloc.ID, task, r.alloc.TaskStates[task.Name], restartTracker)
 		r.tasks[name] = tr
 
 		// Skip tasks in terminal states.
-		if status.Status == structs.AllocClientStatusDead ||
-			status.Status == structs.AllocClientStatusFailed {
+		if state.State == structs.TaskStateDead {
```
Comment (diptanu) [inline client/alloc_runner.go]: Hopefully alloc dirs of Tasks in dead states would be cleaned up without their task runners?

File: client/alloc_runner.go [hunk 3]
```diff
@@ -323,15 +322,15 @@ func (r *AllocRunner) Run() {
 	// Start the task runners
 	r.taskLock.Lock()
 	for _, task := range tg.Tasks {
-		// Skip tasks that were restored
-		if _, ok := r.taskStatus[task.Name]; ok {
+		if _, ok := r.restored[task.Name]; ok {
```
Comment (diptanu) [inline client/alloc_runner.go]: May be in another PR, but it might be nice to not have logic like this in Task Runner where it decides whether to run itself or not?

File: client/alloc_runner.go [hunk 4]
```diff
@@ -344,7 +343,6 @@ OUTER:
 		case update := <-r.updateCh:
 			// Check if we're in a terminal status
 			if update.TerminalStatus() {
-				r.setAlloc(update)
```
Comment (diptanu) [inline client/alloc_runner.go]: Why did we remove this? I think no one else is even calling it or referencing it, probably we should delete the method if this isn't needed anymore?

File: client/task_runner.go [hunk 5]
```diff
@@ -132,12 +137,19 @@ func (r *TaskRunner) DestroyState() error {
 	return os.RemoveAll(r.stateFilePath())
 }
 
-// setStatus is used to update the status of the task runner
-func (r *TaskRunner) setStatus(status, desc string) {
+// setState is used to update the state of the task runner
+func (r *TaskRunner) setState(state string, event *structs.TaskEvent) {
+	// Update the task.
+	r.state.State = state
+	r.state.Events = append(r.state.Events, event)
```
Comment (diptanu) [inline client/task_runner.go]: Consider this to be of fixed length instead of keeping it un-bounded?

File: client/task_runner.go [hunk 6]
```diff
@@ -132,12 +137,19 @@ func (r *TaskRunner) DestroyState() error {
...
+
+	// Persist our state to disk.
 	if err := r.SaveState(); err != nil {
 		r.logger.Printf("[ERR] client: failed to save state of Task Runner: %v", r.task.Name)
 	}
-	r.updater(r.task.Name, status, desc)
+
+	// Indicate the task has been updated.
+	r.updater(r.task.Name)
```
Comment (diptanu) [inline client/task_runner.go]: Not quite sure how the alloc runner is getting the state of the task now.

File: client/task_runner.go [hunk 7]
```diff
@@ -166,94 +179,115 @@ func (r *TaskRunner) startTask() error {
...
 	r.logger.Printf("[DEBUG] client: starting task context for '%s' (alloc '%s')",
 		r.task.Name, r.allocID)
 
-	// Start the task if not yet started
-	if r.handle == nil {
-		if err := r.startTask(); err != nil {
-			return
-		}
-	}
+	r.run(false)
```
Comment (diptanu) [inline client/task_runner.go]: Will the stack ever become very deep if we restart in a hot loop when the user restarts with a very little delay and restart interval and the task segfaults?

**Comment by diptanu** [comment]
@dadgar Overall looks a good step in the right direction with respect to TaskStates, and design looks fine too. Left you some questions.

File: client/alloc_runner.go (hunk 2 above)
Comment (dadgar) [inline client/alloc_runner.go]: Yep it is the alloc_runner that does the destroy.

File: client/alloc_runner.go (hunk 3 above)
Comment (dadgar) [inline client/alloc_runner.go]: Yeah totally agree. There was just a lot already happening in this PR.

File: client/alloc_runner.go (hunk 4 above)
Comment (dadgar) [inline client/alloc_runner.go]: Removed the method as well. I removed it because there was a race between this setting the status and the `dirtySyncState` when the user stops the job. It would trigger an update after the destroy so the incorrect status would be set.

File: client/task_runner.go (hunk 6 above)
Comment (dadgar) [inline client/task_runner.go]: The state of the task is in the alloc within `allocation.TaskStates`

File: client/task_runner.go (hunk 7 above)
Comment (dadgar) [inline client/task_runner.go]: Yeah I will make this a loop. Good call.

File: client/task_runner.go (hunk 5 above)
Comment (dadgar) [inline client/task_runner.go]: Yep lets do it

File: api/tasks.go (hunk 1 above)
Comment (dadgar) [inline api/tasks.go]: I will in setState as suggested below

File: nomad/structs/structs.go [hunk 8]
```diff
@@ -1053,6 +1053,93 @@ func (t *Task) GoString() string {
 	return fmt.Sprintf("*%#v", *t)
 }
 
+// Set of possible states for a task.
+const (
+	TaskStatePending = "pending" // The task is waiting to be run.
+	TaskStateRunning = "running" // The task is currently running.
+	TaskStateDead    = "dead"    // Terminal state of task.
```
Comment (dadgar) [inline nomad/structs/structs.go]: Ah there is a confusion. The State can only be `pending`, `running`, `dead`. However there are a lot of ways a task can enter these states. The `TaskEvent` are those. If you think of it as a finite state machine, the State are the nodes and the Events are the transitions.

File: nomad/structs/structs.go [hunk 9]
```diff
@@ -1053,6 +1053,93 @@ func (t *Task) GoString() string {
...
+}
+
+func (e *TaskEvent) SetDriverError(err error) *TaskEvent {
+	if err != nil {
+		e.DriverError = err.Error()
+	}
+	return e
+}
+
+func (e *TaskEvent) SetExitCode(c int) *TaskEvent {
```
Comment (dadgar) [inline nomad/structs/structs.go]: Nope it is just creating a builder interface which lets you do nice things like this:
`structs.NewTaskEvent(structs.TaskTerminated).SetExitCode(res.ExitCode).SetSignal(res.Signal)`
